src/main/scala/gemmini/TilerFSM.scala

```diff
@@ -0,0 +1,718 @@
+//===========================================================================
+// TilerController's Internal FSM implementation
+//===========================================================================
+package gemmini
+
+import chisel3._
+import chisel3.util._
+import chisel3.experimental._
+import freechips.rocketchip.config._
+import freechips.rocketchip.tile._
+import GemminiISA._
+
+class TilerFSM[T <: Data : Arithmetic, U <: Data]
+  (config: GemminiArrayConfig[T,U])(implicit val p: Parameters) 
+  extends Module with HasCoreParameters {
+  import config._
+
+  //=========================================================================
+  // interface
+  //=========================================================================
+  val io = IO(new Bundle {
+    val cmd_in    = Flipped(Decoupled(new TilerCmd(LOG2_ROB_ENTRIES)))
+    val sched_out = Decoupled(new RoCCCommand)
+    val busy      = Output(Bool())
+  })
+  // hardcode 8 entries with up to 2 pushes per cycle
+  val schedq = Module(new MultiTailedQueue(new RoCCCommand, 8, 2))
+  io.sched_out <> schedq.io.deq
+
+  val sched = schedq.io.enq
+  val cmd   = io.cmd_in.bits
+  val busy  = io.busy
+
+  // initialize ports/pins
+  for (i <- 0 to 1) {
+    sched.bits(i) := DontCare
+    sched.bits(i).status := io.cmd_in.bits.status
+  }
+  sched.push := 0.U
+  io.cmd_in.ready := false.B
+  busy := true.B
+
+  //=========================================================================
+  // FSM states (see diagram for what each state does)
+  //=========================================================================
+  val (s_IDLE ::
+      s_RESET_OUTPUT_GROUP ::
+      s_RESET_A_TILE_SUBCOL ::
+      s_MOVE_FIRST_B_TILE_INTO_SP ::
+      s_RESET_B_TILE_SUBCOL_IN_SUBROW ::
+      s_MAYBE_MOVE_NEXT_B_TILE_INTO_SP ::
+      s_RESET_A_TILE_SUBROW_IN_SUBCOL ::
+      s_MAYBE_MOVE_A_TILE_INTO_SP ::
+      s_MAYBE_MOVE_D_TILE_INTO_ACC ::
+      s_PRELOAD_B_TILE_INTO_ARRAY_AND_SET_C_ADDR_IN_ACC ::
+      s_DO_MATMUL ::
+      s_MAYBE_MOVE_C_TILE_INTO_MEM ::
+      s_NEXT_A_TILE_SUBROW_IN_SUBCOL ::
+      s_NEXT_B_TILE_SUBCOL_IN_SUBROW ::
+      s_NEXT_A_TILE_SUBCOL ::
+      s_NEXT_OUTPUT_GROUP ::
+      Nil) = Enum(16)
+
+  val state = RegInit(s_IDLE)
+
+  //=========================================================================
+  // Internal State
+  //=========================================================================
+
+  // creates a Reg and the next-state Wire, and returns both
+  def regwire(bits: Int) = {
```

Comment:
Can you add this to the `Util.scala` file instead? I think it would fit better there

Reply:
yes